### PR TITLE
Collision

### PR DESCRIPTION
**Pull request by seanmtracey**
Implementation of #43.

### CODE REVIEW

File: client/viewer/js/main.js [hunk 1]
```diff
@@ -36,6 +36,22 @@ socket.on('update', function(data){
 	viewer.update(data);
 });
 
+socket.on('reassign', function(data){
+
+	console.log("Reassign", data);
```
Comment (JakeChampion) [inline client/viewer/js/main.js]: Can we use the debug module with the namespace beginning with `screens:`. The reason being `console.log` is a blocking call and always logs to `stdout`, we may only want to listen to logs from a specific module.

File: client/viewer/js/main.js [hunk 2]
```diff
@@ -36,6 +36,22 @@ socket.on('update', function(data){
...
+
+	var currentData = viewer.getData();
```
Comment (JakeChampion) [inline client/viewer/js/main.js]: `const`

File: client/viewer/js/main.js [hunk 3]
```diff
@@ -36,6 +36,22 @@ socket.on('update', function(data){
...
+
+	if(currentData.id === data.id && currentData.idUpdated === data.idUpdated){
+
+		var n = Object.assign({}, currentData);
```
Comment (JakeChampion) [inline client/viewer/js/main.js]: `const`

File: client/viewer/js/main.js [hunk 4]
```diff
@@ -36,6 +36,22 @@ socket.on('update', function(data){
...
+			n.id = data.newID
```
Comment (JakeChampion) [inline client/viewer/js/main.js]: indentation is a bit funky

File: client/viewer/js/main.js (hunk 3 above)
Comment (JakeChampion) [inline client/viewer/js/main.js]: Maybe a better variable name?

File: client/viewer/js/viewer.js [hunk 5]
```diff
@@ -17,14 +17,20 @@ const __client = (function(moment) {
...
 		data = newdata;
 		// If ID of this screen has changed, update the UI
 		if (newdata.id && newdata.id !== olddata.id) {
+			newdata.idUpdated = Date.now();
 			populateIDs();
 			updateTitle();
 		}
 
+		if(!olddata.idUpdated && !newdata.idUpdated){
+			newdata.idUpdated = Date.now()
```
Comment (JakeChampion) [inline client/viewer/js/viewer.js]: semicolon

File: server/screens.js [hunk 6]
```diff
@@ -33,6 +34,58 @@ function generateAdminUpdate(sock) {
 	});
 }
 
+function decideWhichScreenGetsToKeepAnID(screenA, screenB){
+
+	// Emit an event to the screens to reassign using the ID and the timestamp 
+	// of when that id was assigned as the identifier for the screen that needs
+	// to reassign. The original screen (and the rest) can ignore this message.
+	const screenToChange = screenA.idUpdated > screenB.idUpdated ? screenA : screenB ;
+	app.io.of('/screens').emit('reassign', { id : screenToChange.id, idUpdated : screenToChange.idUpdated, newID : generateID() });
```
Comment (JakeChampion) [inline server/screens.js]: Break this onto a couple of lines

File: server/screens.js [hunk 7]
```diff
@@ -33,6 +34,58 @@ function generateAdminUpdate(sock) {
...
+
+}
+
+
+function checkForConflictingId(id){
+
+	return assignedIDs.some(existingScreen => {
+		return existingScreen.id == id;
```
Comment (JakeChampion) [inline server/screens.js]: `===`

File: client/viewer/js/main.js (hunk 1 above)
Comment (seanmtracey) [inline client/viewer/js/main.js]: Yup, will do.

File: client/viewer/js/main.js (hunk 3 above)
Comment (seanmtracey) [inline client/viewer/js/main.js]: Agreed. It's a hangover from debugging.

File: server/screens.js [hunk 8]
```diff
@@ -33,6 +34,58 @@ function generateAdminUpdate(sock) {
...
+	});
+
+}
+
+function checkForConflictingScreens(data){
+
+	return assignedIDs.some(existingScreen => {
```
Comment (JakeChampion) [inline server/screens.js]:
```
return assignedIDs.some(existingScreen => {
	return existingScreen.id === data.id && existingScreen.idUpdated !== data.idUpdated;
});
```

File: server/screens.js [hunk 9]
```diff
@@ -33,6 +34,58 @@ function generateAdminUpdate(sock) {
...
+			return false;
+		}
+
+	});
+
+}
+
+function generateID(){
+
+	var newID = parseInt(Math.random() * 99999 | 0, 10);
```
Comment (JakeChampion) [inline server/screens.js]: `let`

What does the bitwise OR operator do here?

File: server/screens.js [hunk 10]
```diff
@@ -56,9 +109,17 @@ module.exports.add = function(socket) {
...
 		// If screen has not cited a specific ID, assign one
 		if (!data.id || !parseInt(data.id, 10)) {
-			data.id = Math.random() * 99999 | 0;
+			data.id = generateID();
+		}
+
+		const thereIsAConflict = checkForConflictingScreens(data);
+
+		if(thereIsAConflict){
+			decideWhichScreenGetsToKeepAnID(data, assignedIDs.filter(s => { return s.id == data.id; } )[0] );
```
Comment (JakeChampion) [inline server/screens.js]: We can replace the `filter()[0]` with the [find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) method

File: client/viewer/js/main.js (hunk 1 above)
Comment (JakeChampion) [inline client/viewer/js/main.js]: :+1: apologies for that

File: server/screens.js (hunk 9 above)
Comment (seanmtracey) [inline server/screens.js]: It rounds down to the nearest whole number. We can use Math.floor() if you prefer?

File: client/viewer/js/main.js (hunk 1 above)
Comment (seanmtracey) [inline client/viewer/js/main.js]: No worries, I didn't check where we were.

File: server/screens.js (hunk 8 above)
Comment (seanmtracey) [inline server/screens.js]: This I had originally (or something very close to it), but I struggle to debug it if it went wrong, so I went for the more verbose option.

File: client/viewer/js/main.js [hunk 11]
```diff
@@ -36,6 +36,24 @@ socket.on('update', function(data){
 	viewer.update(data);
 });
 
+socket.on('reassign', function(data){
+
+	console.log("Reassign", data);
```
Comment (JakeChampion) [inline client/viewer/js/main.js:41]: Note: We should probably not log out to the console in production.

**Comment by JakeChampion** [comment]
Tests needed

**Pull request merged**